src/SymfonyCache/TagSubscriber.php

```diff
@@ -0,0 +1,150 @@
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use FOS\HttpCache\SymfonyCache\Tag\ManagerInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\HeaderBag;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * Adds tag invalidation capabilities to the Symfony HTTP cache
+ *
+ * @author Daniel Leech <daniel@dantleech.com>
+ * 
+ * {@inheritdoc}
+ */
+class TagSubscriber implements EventSubscriberInterface
+{
+    /**
+     * HTTP method to use for remote tag invalidation.
+     */
+    const METHOD_INVALIDATE = 'INVALIDATE';
+    
+    /**
+     * Name for HTTP header containing the tags (for both invalidation and
+     * initial tagging).
+     */
+    const HEADER_TAGS = 'X-TaggedCache-Tags';
+
+    /**
+     * Header which should contain the content digest produced by the Symfony
+     * HTTP cache.
+     */
+    const HEADER_CONTENT_DIGEST = 'X-Content-Digest';
+
+    /**
+     * @var ManagerInterface
+     */
+    private $tagManager;
+
+    /**
+     * @param ManagerInterface $tagManager
+     * @param mixed $options
+     */
+    public function __construct(ManagerInterface $tagManager)
+    {
+        $this->tagManager = $tagManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            Events::PRE_HANDLE => 'preHandle',
+            Events::POST_HANDLE => 'postHandle',
+        ];
+    }
+
+    /**
+     * Check to see if the request is an invalidation request, if so
+     * handle the invalidation.
+     *
+     * @param CacheEvent $event
+     */
+    public function preHandle(CacheEvent $event)
+    {
+        $request = $event->getRequest();
+
+        if (self::METHOD_INVALIDATE!== $request->getMethod()) {
+            return;
+        }
+
+        $event->setResponse(
+            $this->handleInvalidate($request)
+        );
+    }
+
+    /**
+     * Check to see if the response contains tags which should be associated
+     * with the cached page.
+     *
+     * @param CacheEvent
+     */
+    public function postHandle(CacheEvent $event)
+    {
+        $response = $event->getResponse();
+
+        if (false === $response->headers->has(self::HEADER_TAGS)) {
+            return;
+        }
+
+        $this->handleTags($response);
+    }
+
+    private function handleInvalidate(Request $request)
+    {
+        $tags = $this->getTagsFromHeaders($request->headers);
+        $nbCacheEntries = $this->tagManager->invalidateTags($tags);
+
+        return new JsonResponse(array(
+            'Status' => 'PURGED',
+            'NbCacheEntries' => $nbCacheEntries
+        ));
+    }
+
+    private function handleTags(Response $response)
+    {
+        if (!$response->headers->has(self::HEADER_CONTENT_DIGEST)) {
+            throw new \RuntimeException(sprintf(
```

Comment:
Not really sure under which cicumstances the content digest header is sent -- we need it in order to link to the cache, and it seems to always be sent when the HTTP cache stores the page.